File: src/main/kotlin/com/example/uberv/coroutines/JavaInterop.java
package com.example.uberv.coroutines;

import java.util.concurrent.CompletableFuture;

public class JavaInterop {

    public static void main(String[] args) {

    }

    CompletableFuture<String> loadGreetings() {
        CompletableFuture<String> future1 = loadGreeting();
        CompletableFuture<String> future2 = loadName();
        return future1.thenCompose(greeting ->
                future2.thenCompose(name ->
                        CompletableFuture.supplyAsync(() ->
                                greeting + ", " + name + "!")));
    }

    static public CompletableFuture<String> loadGreeting() {
        return CompletableFuture.supplyAsync(() -> "Hello");
    }

    static public CompletableFuture<String> loadName() {
        return CompletableFuture.supplyAsync(() -> System.getProperty("user.name"));
    }
}

File: src/main/kotlin/com/example/uberv/coroutines/JavaInterop.kt
package com.example.learningkotlin.coroutines

import com.example.uberv.coroutines.JavaInterop

fun main(args: Array<String>) {

}

fun dosmth() {
    val future1 = JavaInterop.loadName()
    val future2 = JavaInterop.loadGreeting()
}